drivers: flash: nrf_mram: Add more robustness for write

rghaddab opened the pull request:
When writes are issued to the mram controller without verifying that it is ready, it can cause the reads to be stalled until all writes finish. Fix this by adding a wait busy loop before any MRAM_WORD aligned write.
Moreover when writing to weak words, it can cause a Bus Fault. Avoid this by using the MVDMA to capture the fault and recover from it.

rghaddab committed:
drivers: flash: nrf_mram: wait until mramc is ready before write

When writes are issued to the mram controller without verifying that it
is ready, it can cause the reads to be stalled until all writes finish.
Fix this by adding a wait busy loop before each MRAM_WORD aligned write
and disable autopowerdown before starting the write.

Signed-off-by: Riadh Ghaddab <riadh.ghaddab@nordicsemi.no>

David-Garcia-Polo commented on drivers/flash/soc_flash_nrf_mram.c:
```diff
@@ -152,8 +201,8 @@ static const struct flash_parameters *nrf_mram_get_parameters(const struct devic
 		.caps = {
-			.no_explicit_erase = true,
-		},
+				.no_explicit_erase = true,
```
Is this change desired?

karstenkoenig commented on drivers/flash/soc_flash_nrf_mram.c:
```diff
@@ -111,8 +132,24 @@ static int nrf_mram_write(const struct device *dev, off_t offset, const void *da
-	memcpy((void *)addr, data, len);
+	mram_no_latency_sync_request();
+	for (uint32_t i = 0; i < (len / MRAM_WORD_SIZE); i++) {
+		while (!nrf_mram_ready(addr + (i * MRAM_WORD_SIZE))) {
```
This could result in a non-returning loop on old ironside bundles where the mram isn't forced on by the `mram_no_latency_sync_request()` call. Can we have this fault on timeout or something to make the code reasonably compatible on old variants of the bundle>

hakonfam reviewed (changes requested):
nack to block while looking at issues with compatibility ref @karstenkoenig comment

rghaddab committed:
drivers: nrf_mram: verify writes using MVDMA

Fix the case where the MRAM writes get corrupted in some conditions.
After each "MRAM_WORD_SIZE write", a read of the written/erased MRAM word
is performed using the MVDMA.
If an error is detected in the MVDMA registers, retry the write for a
maximum of CONFIG_NRF_MRAM_MAX_RETRIES (default 5) times.
Add a warning as well in case the write-block-size is not a multiple of
the MRAM_WORD_SIZE and in case of partial writes, read all the unwritten
data and commit it again instead of only the last bytes. This will avoid
data corruption in case of failed partial writes.

Signed-off-by: Riadh Ghaddab <riadh.ghaddab@nordicsemi.no>
